index.js

```diff
@@ -50,103 +50,100 @@ Limiter.prototype.inspect = function () {
  * that expire after N seconds:
  *
  *  - limit:<id>:count
- *  - limit:<id>:limit
- *  - limit:<id>:reset
  *
  * @param {Function} fn
  * @api public
  */
 
 Limiter.prototype.get = function (fn) {
   var count = this.prefix + 'count';
-  var limit = this.prefix + 'limit';
-  var reset = this.prefix + 'reset';
   var duration = this.duration;
   var max = this.max;
   var db = this.db;
 
   function create() {
-    var ex = (Date.now() + duration) / 1000 | 0;
-
     db.multi()
-      .set([count, max, 'PX', duration, 'NX'])
-      .set([limit, max, 'PX', duration, 'NX'])
-      .set([reset, ex, 'PX', duration, 'NX'])
+      .set(count, max, 'PX', duration, 'NX')
       .exec(function (err, res) {
         if (err) return fn(err);
 
         // If the request has failed, it means the values already
         // exist in which case we need to get the latest values.
-        if (isFirstReplyNull(res)) return mget();
-
-        fn(null, {
-          total: max,
-          remaining: max,
-          reset: ex
+        if (isReplyNull(res)) return get();
+
+        db.pttl(count, function(err, data) {
+          if(err) return fn(err);
+          fn(null, {
+            total: max,
+            remaining: max,
+            // Adding Date.now to ensure change is non-breaking. Can be removed for more accuracy
+            // later. reset should also be in milliseconds (unlike now) for more accuracy
+            reset: (~~data + Date.now()) / 1000
+          });
         });
       });
   }
 
   function decr(res) {
-    var n = ~~res[0];
-    var max = ~~res[1];
-    var ex = ~~res[2];
-    var dateNow = Date.now();
+    var n = ~~res;
 
     if (n <= 0) return done();
 
     function done() {
-      fn(null, {
-        total: max,
-        remaining: n < 0 ? 0 : n,
-        reset: ex
+      db.pttl(count, function(err, data) {
+        if(err) return fn(err);
+        fn(null, {
+          total: max,
+          remaining: n < 0 ? 0 : n,
+          // Adding Date.now to ensure change is non-breaking. Can be removed for more accuracy
+          // later. reset should also be in milliseconds (unlike now) for more accuracy
+          reset: (~~data + Date.now()) / 1000
+        });
       });
     }
 
     db.multi()
-      .set([count, n - 1, 'PX', ex * 1000 - dateNow, 'XX'])
-      .pexpire([limit, ex * 1000 - dateNow])
-      .pexpire([reset, ex * 1000 - dateNow])
+      .decr(count)
```

Comment:
1000 calls sometimes are not enough to simulate race condition. Try to think about that, 2 simultaneous calls to 2 different servers processes, the count is 1. The two calls pass the condition, now both of them are going to decrease the value, and set the value to -1.